firedrake/mg/embedded.py

```diff
@@ -173,11 +173,16 @@ def DG_work(self, V):
         :arg V: a function space.
         :returns: A Function in the embedding DG space.
         """
+        needs_dual = ufl.duals.is_dual(V)
         cache = self.cache(V.ufl_element())
-        key = V.dim()
+        key = (V.dim(), needs_dual)
```

Comment:
@dham This should address point (1/2) from the review. I used recursion to populate both Function and Cofunction caches.